shim/src/handle_vcpu_kvm_get_fpu.c

```diff
@@ -24,20 +24,54 @@
  * SOFTWARE.
  */
 
+#include <debug.h>
+#include <detect_hypervisor.h>
+#include <g_mut_hndl.h>
 #include <kvm_fpu.h>
+#include <mv_constants.h>
+#include <mv_fpu_state_t.h>
+#include <mv_hypercall.h>
 #include <mv_types.h>
+#include <platform.h>
+#include <shared_page_for_current_pp.h>
+#include <shim_vcpu_t.h>
 
 /**
  * <!-- description -->
  *   @brief Handles the execution of kvm_get_fpu.
  *
  * <!-- inputs/outputs -->
  *   @param pmut_ioctl_args the arguments provided by userspace
+ *   @param  vcpu arguments received from private data
  *   @return SHIM_SUCCESS on success, SHIM_FAILURE on failure.
  */
 NODISCARD int64_t
-handle_vcpu_kvm_get_fpu(struct kvm_fpu *const pmut_ioctl_args) NOEXCEPT
+handle_vcpu_kvm_get_fpu(
+    struct shim_vcpu_t const *const vcpu, struct kvm_fpu *const pmut_ioctl_args) NOEXCEPT
 {
-    (void)pmut_ioctl_args;
+    struct mv_fpu_state_t *pmut_mut_fpu;
+
+    platform_expects(MV_INVALID_HANDLE != g_mut_hndl);
+    platform_expects(NULL != vcpu);
+    platform_expects(NULL != pmut_ioctl_args);
+
+    if (detect_hypervisor()) {
+        bferror("The shim is not running in a VM. Did you forget to start MicroV?");
+        return SHIM_FAILURE;
+    }
+
+    pmut_mut_fpu = (struct mv_fpu_state_t *)shared_page_for_current_pp();
+    platform_expects(NULL != pmut_mut_fpu);
+
+    if (mv_vs_op_fpu_get_all(g_mut_hndl, vcpu->vsid)) {
+        bferror("mv_vs_op_reg_get_all failed");
+        return SHIM_FAILURE;
+    }
+
+    platform_memcpy(pmut_ioctl_args->registers, pmut_mut_fpu->registers, NO_OF_REGISTERS_BYTES);
```

Comment:
Maybe break this copying out into a function so it can be reused with set_fpu below? If there isn't an obvious header that is shared by both then feel free to ignore this comment.

Reply:
we have a header file which is related to that specific ioctl.

Reply:
But not one shared between get_fpu and set_fpu right? Seems as though we'll have to live with a decent amount of duplication with how things are structured at the moment.